vyper/ast/README.md

```diff
@@ -0,0 +1,127 @@
+# 🐍 `vyper.ast` 🐍
+
+## Purpose
+
+The `vyper.ast` module handles the conversion of Vyper source code into an abstract
+syntax tree (AST). It also contains functionality for traversing and modifying the AST,
+and parsing NatSpec docstrings.
+
+## Organization
+
+`vyper.ast` has the following structure:
+
+* [`annotation.py`](annotation.py): Contains the `AnnotatingVisitor` class, used to
+annotate and modify the Python AST prior to converting it to a Vyper AST.
+* [`natspec.py`](natspec.py): Functions for parsing NatSpec docstrings within the
+source.
+* [`nodes.py`](nodes.py): Contains the Vyper node classes, and the `get_node`
+function which generates a Vyper node from a Python node.
+* [`pre_parser.py`](pre_parser.py): Functions for converting Vyper source into
+parseable Python source.
+* [`utils.py`](utils.py): High-level functions for converting source code into AST
+nodes.
+
+## Control Flow
+
+### Node Generation
+
+`vyper.ast.utils.parse_to_ast` is the main function used to generate a Vyper AST
+from a source string. The process is as follows:
+
+1. In [`pre_parser.py`](pre_parser.py), Vyper source is modified into parseable
+Python. This primarily involves substituting out `contract` and `struct` statements
+for `class`.
+2. A Python AST is generated from the modified source.
+3. In [`annotation.py`](annotation.py), additional information is added to the
+Python AST, and some nodes are modified to aid conversion to the Vyper AST.
+4. In [`nodes.py`](nodes.py), the Python AST nodes are converted to Vyper
+AST nodes.
+
+Conversion between a Python Node and a Vyper Node uses the following rules:
+
+* The type of Vyper node is determined from the `ast_type` field of the Python node.
+* Fields listed in `__slots__` may be included and may have a value.
+* Fields listed in `_translated_fields` have their key modified prior to being added.
+This is used to handle discrepencies in how nodes are structured between different
+Python versions.
+* Fields listed in `_only_empty_fields`, if present within the Python AST, must
+be `None` or a `SyntaxException` is raised.
+* All other fields are ignored.
+
+Most Vyper nodes have an identical (or simplified) structure to their Python
+counterpart. Divergences are always handled in a Vyper node's `__init__` method,
+and explained in the docstring. To learn more about the structure of the nodes:
+
+* Green Tree Snakes: [Meet the Nodes](https://greentreesnakes.readthedocs.io/en/latest/nodes.html)
+
+### Traversing the AST
+
+Each node contains several methods that are used to traverse the AST. These methods
+can optionally include one or more filters in order to quickly search for children
+or parents that match a desired pattern.
+
+* `get_children`: Returns direct childen of the given node.
+* `get_descendants`: Returns direct and indirect childen of the given node.
+* `get_parent`: Returns a parent of the given node.
```

Comment:
"the" because there is only one
```suggestion
* `get_parent`: Returns the parent of the given node.
```

Reply:
I fixed the description but maybe the name of the method is wrong. When a filter is given, it returns the first parent, grandparent, great-great-great-granddaddy, etc. that matches the filter.  Perhaps `get_ascendant` or `get_ancestor` would be clearer?

Reply:
`ancestor`